test/Microsoft.AspNet.Mvc.Core.Test/ViewDataDictionaryTest.cs

```diff
@@ -341,33 +329,57 @@ public void CopyConstructors_CopyModelMetadata_ForTypeObject()
             // Assert
             Assert.NotNull(viewData1.ModelMetadata);
             Assert.Equal(expectedType, viewData1.ModelMetadata.ModelType);
-            Assert.Equal(expectedType, viewData1.ModelMetadata.RealModelType);
 
             Assert.NotNull(viewData2.ModelMetadata);
             Assert.Equal(expectedType, viewData2.ModelMetadata.ModelType);
-            Assert.Equal(expectedType, viewData2.ModelMetadata.RealModelType);
         }
 
         [Fact]
-        public void ModelSetter_UpdatesModelMetadata()
+        public void ModelSetter_SameType_UpdatesModelExplorer()
         {
             // Arrange
             var metadataProvider = new EmptyModelMetadataProvider();
             var viewData = new ViewDataDictionary(metadataProvider)
             {
-                Model = "same-string",
+                Model = 3,
             };
+
+            var originalMetadata = viewData.ModelMetadata;
+            var originalExplorer = viewData.ModelExplorer;
+
+            // Act
+            viewData.Model = 5;
+
+            // Assert
+            Assert.NotNull(viewData.ModelMetadata);
+            Assert.Equal(5, viewData.Model);
+            Assert.Same(originalMetadata, viewData.ModelMetadata);
+            Assert.NotSame(originalExplorer, viewData.ModelExplorer);
```

Comment:
nit: suggest adding a unit test showing `Assert.NotSame(originalExplorer, viewData.ModelExplorer);` even if line 151 were `viewData.Model = 3;` (same value type in a different box).